### PR TITLE
SC: Add checking the already handled events for Value Transfer Recovery

### PR DESCRIPTION
**Pull request by KimKyungup**
## Proposed changes

This PR added the routine checking the already handled event during value transfer recovery.
This makes VTR (Value Transfer Recovery) to be faster. And If the 1000 events is already handled from upperHandleNonce, this logic can deal with this situation.

## Types of changes

Please put an x in the boxes related to your change.

- [ ] Bugfix
- [x] New feature or enhancement
- [ ] Others

## Checklist

*Put an x in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of them, don't hesitate to ask. We're here to help! This is simply a reminder of what we are going to look for before merging your code.*

- [x] I have read the [CONTRIBUTING GUIDELINES](https://github.com/klaytn/klaytn/blob/master/CONTRIBUTING.md) doc
- [x] I have signed the [CLA](https://cla-assistant.io/klaytn/klaytn)
- [x] Lint and unit tests pass locally with my changes (`$ make test`)
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added necessary documentation (if appropriate)
- [ ] Any dependent changes have been merged and published in downstream modules

## Related issues

- Please leave the issue numbers or links related to this PR here.

## Further comments

If this is a relatively large or complex change, kick off the discussion by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

File: node/sc/vt_recovery.go [hunk 1]
```diff
@@ -271,13 +271,20 @@ pendingTxLoop:
...
+		it, err := from.bridge.FilterRequestValueTransfer(opts, nil, nil, nil)
 		if err != nil {
 			return nil, err
 		}
 		for it.Next() {
 			logger.Trace("pending nonce in the event", "requestNonce", it.Event.RequestNonce)
 			if it.Event.RequestNonce >= hint.handleNonce {
+				// Check if the event is already handled in target bridge contract
+				blk, err := to.bridge.HandleNoncesToBlockNums(nil, it.Event.RequestNonce)
+				if err == nil && blk > 0 {
```
Comment (jeonghonah) [inline node/sc/vt_recovery.go]: Typically, recovery will operate in an abnormal situation and will have to be treated as a batch, but looking up a contact at each event may be a performance disadvantage.

**Comment by KimKyungup** [comment]
> LGTM.
> 
> It will be nice to check performance results by using some micro benchmarks.

@jeonghonah I tested the performance test.

I measured the time duration for requesting about 20,000 requests.
And After that I stopped locustSlave and rebooted the KSEN and I measured the time duration recovery all requests.

**time duration to receive 20,000 requests**

Before | After 
-------|------ 
214s | 134s 

**time duration to recover all remain requests**

Before | After
-------|------ 
162s | 101s


This is raw data.
```
INFO[11/26,09:11:42 +09] [45] VT : Parent -> Child Chain                request=0 handle=0 lowerHandle=0 pending=0
INFO[11/26,09:11:42 +09] [45] VT : Child -> Parent Chain                request=0 handle=0 lowerHandle=0 pending=0
WARN[11/26,09:11:42 +09] [45] Unregistered counter part token address.  addr=0x0000000000000000000000000000000000000000
INFO[11/26,09:11:42 +09] [45] Register counter part token address.      addr=0xcEc1b246901b96193b60a3767e50efCEfED06443 cpAddr=0xcEc1b246901b96193b60a3767e50efCEfED06443
WARN[11/26,09:11:42 +09] [45] Unregistered counter part token address.  addr=0x0000000000000000000000000000000000000000
INFO[11/26,09:11:42 +09] [45] Register counter part token address.      addr=0x0EE9c98C6159Ce77a5012dD429517291A8e886E2 


INFO[11/26,09:15:16 +09] [45] VT : Parent -> Child Chain                request=10309 handle=5958 lowerHandle=5958 pending=4351
INFO[11/26,09:15:16 +09] [45] VT : Child -> Parent Chain                request=9819  handle=6551 lowerHandle=6551 pending=3268
——————————————————
15:16 - 11:42 = 3:34


INFO[11/26,09:15:49 +09] [45] VT : Parent -> Child Chain                request=6175 handle=6436 lowerHandle=6436 pending=18446744073709551355
INFO[11/26,09:15:49 +09] [45] VT : Child -> Parent Chain                request=6670 handle=6670 lowerHandle=6670 pending=0
INFO[11/26,09:15:49 +09] [45] retrieved pending events                  bridge=0x287cFfb3BfbEe35a816910d625cb2c12Fb8BDE9D len(pendingEvents)=1000 1st nonce=6175
WARN[11/26,09:15:49 +09] [45] VT Recovery : Child -> Parent Chain       cBridge=0x07d53893a7B0a51b1d9e2e6aD8a926F490716cDb events=1000
WARN[11/26,09:15:49 +09] [45] VT Recovery : Parent -> Child Chain       pBridge=0x287cFfb3BfbEe35a816910d625cb2c12Fb8BDE9D events=1000


INFO[11/26,09:18:31 +09] [45] VT : Parent -> Child Chain                request=10626 handle=10626 lowerHandle=10626 pending=0
INFO[11/26,09:18:31 +09] [45] VT : Child -> Parent Chain                request=10465 handle=10465 lowerHandle=10465 pending=0


18:31 - 15:49 =  2:42
120


———————————————————————
INFO[11/26,09:40:04 +09] [45] VT : Parent -> Child Chain                request=0 handle=0 lowerHandle=0 pending=0
INFO[11/26,09:40:04 +09] [45] VT : Child -> Parent Chain                request=0 handle=0 lowerHandle=0 pending=0
WARN[11/26,09:40:04 +09] [45] Unregistered counter part token address.  addr=0x0000000000000000000000000000000000000000
INFO[11/26,09:40:04 +09] [45] Register counter part token address.      addr=0x9dC6c4BBd4cb22197d499ECf15aC0796dC948df6 cpAddr=0x9dC6c4BBd4cb22197d499ECf15aC0796dC948df6
WARN[11/26,09:40:04 +09] [45] Unregistered counter part token address.  addr=0x0000000000000000000000000000000000000000
INFO[11/26,09:40:04 +09] [45] Register counter part token address.      addr=0x5732035561fEd5f7B72C8f21E29270f2D374cA7B cpAddr=0x5732035561fEd5f7B72C8f21E29270f2D374cA7B
INFO[11/26,09:40:05 +09] [5] Imported new chain segment                number=238 hash=7f3709…431d34 blocks=1  txs=40  elapsed=65.191ms  trieDBSize=13.58mB  



INFO[11/26,09:42:18 +09] [45] VT : Parent -> Child Chain                request=10012 handle=2215 lowerHandle=2215 pending=7797
INFO[11/26,09:42:18 +09] [45] VT : Child -> Parent Chain                request=9976  handle=2557 lowerHandle=2557 pending=7419
——————————————————

42:18 - 40:04 = 2:14


INFO[11/26,09:43:16 +09] [45] retrieved pending events                  bridge=0xAc260f1a828710A844825945e0Cb83ba4BC15E7e len(pendingEvents)=1000 1st nonce=2957
INFO[11/26,09:43:16 +09] [47] Persisted trie from memory database       updated nodes=135 updated nodes size=18.07kB time=5.323214ms gcnodes=0 gcsize=0.00B gctime=4.831µs livenodes=1 livesize=0.00B
INFO[11/26,09:43:16 +09] [5] Imported new chain segment                number=385 hash=8eb757…d287c2 blocks=13 txs=453 elapsed=1.194s trieDBSize=78.70kB mgas=31.480 mgasps=26.364
INFO[11/26,09:43:17 +09] [45] VT : Parent -> Child Chain                request=10606 handle=2230 lowerHandle=2230 pending=8376
INFO[11/26,09:43:17 +09] [45] VT : Child -> Parent Chain                request=10063 handle=3116 lowerHandle=3116 pending=6947



INFO[11/26,09:44:57 +09] [45] VT : Parent -> Child Chain                request=10606 handle=10512 lowerHandle=10512 pending=94
INFO[11/26,09:44:57 +09] [45] VT : Child -> Parent Chain                request=10424 handle=10424 lowerHandle=10424 pending=0
INFO[11/26,09:44:58 +09] [5] Imported new chain segment                number=493 hash=b73a43…ee3867 blocks=1  txs=66   elapsed=20.067ms  trieDBSize=2.37mB   mgas=1.915  mgasps=95.451
INFO[11/26,09:44:58 +09] [45] VT : Parent -> Child Chain                request=10606 handle=10606 lowerHandle=10606 pending=0
INFO[11/26,09:44:58 +09] [45] VT : Child -> Parent Chain                request=10424 handle=10424 lowerHandle=10424 pending=0
———
44:58 - 43:17 = 1:41
```

File: node/sc/vt_recovery.go [hunk 2]
```diff
@@ -244,8 +244,8 @@ func (vtr *valueTransferRecovery) retrievePendingEvents() error {
 
 // retrievePendingEventsFrom retrieves pending events from the specified bridge by using the hint provided.
 // The filter uses a hint as a search range. It returns a slice of events that has log details.
-func retrievePendingEventsFrom(hint *valueTransferHint, bi *BridgeInfo) ([]*RequestValueTransferEvent, error) {
-	if bi.bridge == nil {
+func retrievePendingEventsFrom(hint *valueTransferHint, from, to *BridgeInfo) ([]*RequestValueTransferEvent, error) {
+	if from.bridge == nil {
```
Comment (ehnuje) [inline node/sc/vt_recovery.go]: Don't we need to check if `to.bridge` is nil or not?

Comment (KimKyungup) [reply node/sc/vt_recovery.go]: Good point. I reflected.

**Comment by KimKyungup** [comment]
@ehnuje @jeonghonah PTAL. :)

**Pull request merged**